storage/StorageExample/Shared/ContentView.swift

```diff
@@ -0,0 +1,261 @@
+        }
+      }
+      .onChange(of: vm.inputImage) { _ in
+        loadImage()
+      }
+      .onAppear {
+        firebaseAuth()
+      }
+      .alert("Error", isPresented: $vm.errorFound) {
+        Text(vm.errInfo.debugDescription)
+        Button("ok") {}
+      }
+      .alert("Image was uploaded", isPresented: $vm.fileUploaded) {
+        Button("ok") {}
+        Button("Link") {
+          if let url = vm.fileDownloadURL {
+            print("downloaded url: \(url)")
+            UIApplication.shared.open(url)
+          }
+        }
+      }
+
+      if vm.isLoading {
+        LoadingView()
+      }
+    }
+  }
+
+  func loadImage() {
+    guard let inputImage = vm.inputImage else {
+      return
+    }
+    vm.image = Image(uiImage: inputImage)
+  }
+
+  func uploadFromALocalFile() {
+    let filePath = Auth.auth().currentUser!.uid +
+      "/\(Int(Date.timeIntervalSinceReferenceDate * 1000))/\(imageURL.lastPathComponent)"
+    // [START uploadimage]
+    let storageRef = storage.reference(withPath: filePath)
+
+    vm.isLoading = true
+    storageRef.putFile(from: imageURL, metadata: nil) { metadata, error in
+      guard let _ = metadata else {
+        // Uh-oh, an error occurred!
+        vm.errorFound = true
+        vm.errInfo = error
+        return
+      }
+
+      vm.isLoading = false
+      // You can also access to download URL after upload.
+      fetchDownloadURL(storageRef, storagePath: filePath)
+    }
+  }
+
+  func uploadFromData() {
+    guard let imageData = vm.inputImage?.jpeg else {
+      print("The image from url \(imageURL.path) cannot be transferred to data.")
+      return
+    }
+    let filePath = Auth.auth().currentUser!.uid +
+      "/\(Int(Date.timeIntervalSinceReferenceDate * 1000))/fromData/\(imageURL.lastPathComponent)"
+    // [START uploadimage]
+    let storageRef = storage.reference(withPath: filePath)
+
+    vm.isLoading = true
+    storageRef.putData(imageData, metadata: nil) { metadata, error in
+      guard let _ = metadata else {
+        // Uh-oh, an error occurred!
+        vm.errorFound = true
+        vm.errInfo = error
+        return
+      }
+      vm.isLoading = false
+      // You can also access to download URL after upload.
+      fetchDownloadURL(storageRef, storagePath: filePath)
+    }
+  }
+
+  func fetchDownloadURL(_ storageRef: StorageReference, storagePath: String) {
+    storageRef.downloadURL { url, error in
+
+      guard let downloadURL = url else {
+        print("Error getting download URL: \(error.debugDescription)")
+        vm.errorFound = true
+        self.vm.errInfo = error
+        return
+      }
+      print("download url: \(downloadURL.absoluteString)")
+      UserDefaults.standard.set(storagePath, forKey: "storagePath")
+      UserDefaults.standard.synchronize()
+      vm.downloadPicButtonEnabled = true
+      vm.fileUploaded = true
+      vm.fileDownloadURL = downloadURL
+    }
+  }
+
+  func download() {
```

Comment:
nit: could you make the API more descriptive? Maybe?:
```suggestion
  func downloadImage() {
```